### ISSUE TITLE
[JAVA] enhance beanvalidation with adding @Valid

### ISSUE THREAD

**Issue by fisherliu** [swagger-api/swagger-codegen#5498]
hi @wing328 @fehguy @xhh ,

swagger-codegen-plugin is great plugin, we all benefit from it and  love it! Thank you for great your contribution for it.

Could we ask swagger-codegen  to support @Valid annotation for beanValidation to cascade beans-validation recursively? This feature is very useful for the systems.
for example, if  a Project object has a List<Person>, current swagger-codegen 2.3.0 not recursively validates the Person fields info like Person.id which should be not empty

What's more, another concern is that when List is required,  we hope the List size is greater than zero not just empty List. we can add @Size(min = 1) to implement this.
Yaml file:
```
Project:
    type: object
    required:
    - people
    properties:
      people:
        type: array
        items:
          $ref: '#/definitions/Person'
# PERSON
  Person:
    type: object
    description: the info of a person
    required:
    - id
    - name
    properties:
      id:
        type: string
        description: Unique identifier for the person
	  name:
        type: string
        description: The name for the person
```
The generated result with current version:
  ```
@NotNull
  @ApiModelProperty(example = "null", required = true, value = "")
  public List<Person> getPeople() {
    return people;
  }
```
The expected result is like below:
```
  @Valid
  @NotNull
  @Size(min = 1)
  @ApiModelProperty(example = "null", required = true, value = "")
  public List<Person> getPeople() {
    return people;
  }
```

In fact, I made some change in my local env  and implemented this feature.
/swagger-codegen/src/main/resources/Java/beanValidation.mustache 
i added below

```
{{#recursiveBeanValidation}}
{{^isPrimitiveType}}
  @Valid
{{/isPrimitiveType}}
{{/recursiveBeanValidation}}
{{#required}}
  @NotNull
{{#isListContainer}}
  @Size(min = 1)
{{/isListContainer}}
{{#isString}}
 {{^isEnum}}
  @Size(min = 1)
 {{/isEnum}}
{{/isString}}
{{/required}} 
```

/swagger-codegen/src/main/resources/Java/model.mustache 
I added below:

```
{{#recursiveBeanValidation}}
import javax.validation.Valid;
{{/recursiveBeanValidation}} 
```

/swagger-codegen/src/main/java/io/swagger/codegen/languages/JavaClientCodegen.java 
I added recursiveBeanValidation flag to support this feature below like useBeanValidation
`protected boolean recursiveBeanValidation = false;`
`cliOptions.add(CliOption.newBoolean("recursiveBeanValidation", "recursive BeanValidation API annotations"));`

```
if (additionalProperties.containsKey("recursiveBeanValidation")) {
	this.setRecursiveBeanValidation(convertPropertyToBooleanAndWriteBack("recursiveBeanValidation"));

}
```
```
public void setRecursiveBeanValidation(boolean recursiveBeanValidation) {
		this.recursiveBeanValidation = recursiveBeanValidation;
	}
```

when we want to use the feature, we can add recursiveBeanValidation in 
       ` <plugin>`
				`<groupId>io.swagger</groupId>`
				`<artifactId>swagger-codegen-maven-plugin</artifactId>`
				`<version>2.3.0</version>`
				`<executions>`
					`<execution>`
						`<goals>`
							`<goal>generate</goal>`
						`</goals>`
						`<configuration>`
							`<inputSpec>XXXX.yaml</inputSpec>`
							`<language>java</language>`
							`<environmentVariables>`
								`<models></models>`
							`</environmentVariables>`
							`<modelPackage>com.xxxx.model</modelPackage>`
							`<configOptions>`
							    `**<recursiveBeanValidation>true</recursiveBeanValidation>**`
							    `<useBeanValidation>true</useBeanValidation>`
								`<serializableModel>true</serializableModel>`
							`</configOptions>`
						`</configuration>`
					`</execution>`
				`</executions>`
			`</plugin>`

Thank you again!

**Comment by wing328** [comment]
@fisherliu we've the following PRs. Are these what you're looking for?

https://github.com/swagger-api/swagger-codegen/pull/5393
https://github.com/swagger-api/swagger-codegen/pull/5391 (merged)

**Comment by fisherliu** [comment]
Thank you for your timely reply.
Yes, these are what  I am looking for. 
Which version will they be merged to?

**Comment by fisherliu** [comment]
@wing328, another concern is that when the List is required, we hope the List size is greater than zero not just empty List, could we add @Size(min = 1) to implement this.

**Comment by wing328** [comment]
> Which version will they be merged to?

Current master (2.2.3), which will be sync'd to 2.3.0 as well.

> we hope the List size is greater than zero not just empty List, could we add @Size(min = 1) to implement this.

Have you set `minItems` in the spec?

https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#fixed-fields-7

**Comment by fisherliu** [comment]
#5393:
https://github.com/cakemail/swagger-codegen/blob/07f066a20350302429c133f2febfcb58632aa56a/modules/swagger-codegen/src/main/resources/JavaSpring/beanValidation.mustache

the change is added into JavaSpring, will it be added into Java language?

**Comment by fisherliu** [comment]
Thank you for your help.

**Comment by wing328** [comment]
> the change is added into JavaSpring, will it be added into Java language?

I don't think anyone is working on that. Would you have time to contribute the enhancement?

**Comment by fisherliu** [comment]
I don't have the push access.
File uploads require push access to this repository.

remote: Permission to swagger-api/swagger-codegen.git denied to fisherliu.

**Comment by wing328** [comment]
You need to fork this repo to begin with. Please take a look at https://github.com/swagger-api/swagger-codegen/wiki/FAQ#how-can-i-submit-a-pr-to-fix-bugs-or-make-enhancements to see if that helps.

**Comment by fisherliu** [comment]
The changes are like below. I tested simply . Could you or someone else help add?
swagger-codegen\modules\swagger-codegen\src\main\resources\Java\beanValidation.mustache
```
{{#isContainer}}
{{^isPrimitiveType}}
{{^isEnum}}
  @Valid
{{/isEnum}}
{{/isPrimitiveType}}
{{/isContainer}}
{{#isNotContainer}}
{{^isPrimitiveType}}
  @Valid
{{/isPrimitiveType}}
{{/isNotContainer}}
{{>beanValidationCore}}
```
swagger-codegen\modules\swagger-codegen\src\main\resources\Java\model.mustache
```
{{#useBeanValidation}}
import javax.validation.constraints.*;
**import javax.validation.Valid;**
{{/useBeanValidation}}
```

**Comment by wing328** [comment]
Closed via #5517
